File: Advanced/PageAndRDL/CustomPdfExport/C#/PdfRendering/Rendering/PdfImagesFactory.cs
using System;
using System.Collections.Generic;
using System.IO;
using PdfSharp.Drawing;

namespace GrapeCity.ActiveReports.Samples.Export.Rendering.PdfSharp
{
	internal sealed class PdfImagesFactory : IDisposable
	{
		private readonly IDictionary<string, XImage> _images = new Dictionary<string, XImage>();

		public XImage GetPdfImage(string cacheId, Stream content)
		{
			XImage image;
			if (_images.TryGetValue(cacheId, out image))
				return image;
			image = XImage.FromStream(content);
			_images[cacheId] = image;
			return image;
		}

		#region IDisposable implementation

		void IDisposable.Dispose()
		{
			foreach (var image in _images.Values)
				image.Dispose();
			_images.Clear();
		}

		#endregion
	}
}

File: Advanced/PageAndRDL/CustomPdfExport/VB.NET/PdfRendering/Rendering/PdfImagesFactory.vb
Imports System.IO
Imports PdfSharp.Drawing

Friend NotInheritable Class PdfImagesFactory
	Implements IDisposable
	Private ReadOnly _images As IDictionary(Of String, XImage) = New Dictionary(Of String, XImage)()

	Public Function GetPdfImage(cacheId As String, content As Stream) As XImage
		Dim image As XImage = Nothing
		If _images.TryGetValue(cacheId, image) Then
			Return image
		End If
		image = XImage.FromStream(content)
		_images(cacheId) = image
		Return image
	End Function

#Region "IDisposable implementation"

	Sub Dispose() Implements IDisposable.Dispose
		For Each image As XImage In _images.Values
			image.Dispose()
		Next
		_images.Clear()
	End Sub

#End Region
End Class
